libs/geometry/src/TangentSpaceMesh.cpp

```diff
@@ -278,37 +407,85 @@ TangentSpaceMesh& TangentSpaceMesh::operator=(TangentSpaceMesh&& that) noexcept
 }
 
 size_t TangentSpaceMesh::getVertexCount() const noexcept {
-    return 0;
+    return mOutput->vertexCount;
 }
 
-void TangentSpaceMesh::getPositions(float3*, size_t) const {
+#define STRIDE_TO(out, stride) (out = (decltype(out)) (((uint8_t*) out) + stride))
+
+void TangentSpaceMesh::getPositions(float3* positions, size_t stride) const {
     ASSERT_PRECONDITION(mInput->positions, "Must provide input positions");
+    stride = stride ? stride : sizeof(decltype(*positions));
+    for (size_t i = 0; i < mOutput->vertexCount; ++i) {
+        *positions = mOutput->positions[i];
+        STRIDE_TO(positions, stride);
+    }
 }
 
-void TangentSpaceMesh::getUVs(float2*, size_t) const {
+void TangentSpaceMesh::getUVs(float2* uvs, size_t stride) const {
     ASSERT_PRECONDITION(mInput->uvs, "Must provide input UVs");
+    stride = stride ? stride : sizeof(decltype(*uvs));
+    for (size_t i = 0; i < mOutput->vertexCount; ++i) {
+        *uvs = mOutput->uvs[i];
+        STRIDE_TO(uvs, stride);
+    }
 }
 
 size_t TangentSpaceMesh::getTriangleCount() const noexcept {
-    return 0;
+    return mOutput->triangleCount;
 }
 
-void TangentSpaceMesh::getTriangles(uint3*) const {
+void TangentSpaceMesh::getTriangles(uint3* out) const {
     ASSERT_PRECONDITION(mInput->triangles16 || mInput->triangles32, "Must provide input triangles");
+
+    const bool is16 = mOutput->triangles16 != nullptr;
+    const size_t stride = sizeof(decltype(*out));
+    for (size_t i = 0; i < mOutput->triangleCount; ++i) {
+        *out = is16 ? uint3(mOutput->triangles16[i]) : mOutput->triangles32[i];
+        STRIDE_TO(out, stride);
+    }
 }
 
-void TangentSpaceMesh::getTriangles(ushort3*) const {
+void TangentSpaceMesh::getTriangles(ushort3* out) const {
     ASSERT_PRECONDITION(mInput->triangles16 || mInput->triangles32, "Must provide input triangles");
+
+    const bool is16 = mOutput->triangles16 != nullptr;
+    const size_t stride = sizeof(decltype(*out));
+    for (size_t i = 0; i < mOutput->triangleCount; ++i) {
+        if (is16) {
+            *out = mOutput->triangles16[i];
+        } else {
+            const uint3& tri = mOutput->triangles32[i];
+            ASSERT_PRECONDITION(tri.x <= USHRT_MAX &&
+                                tri.y <= USHRT_MAX &&
+                                tri.z <= USHRT_MAX, "Overflow when casting uint3 to ushort3");
+            *out = ushort3(static_cast<uint16_t>(tri.x),
+                    static_cast<uint16_t>(tri.y),
+                    static_cast<uint16_t>(tri.z));
+        }
+        STRIDE_TO(out, stride);
+    }
 }
 
-void TangentSpaceMesh::getQuats(quatf*, size_t) const noexcept {
+#define GET_QUATS_IMPL(out, stride, conversion) \
+    stride = stride ? stride : sizeof(decltype(*out));\
+    for (size_t i = 0; i < mOutput->vertexCount; ++i) {\
+        *out = conversion(mOutput->tangentSpace[i].xyzw);\
```

Comment:
why do you need to specify `.xyzw`? also I would cache `mOutput->tangentSpace` outside of the loop, because otherwise the compile will have to reload `mOutput` at each iteration.

In addition if you used an inline function, you could mark the in and out pointers as `UTILS_RESTRICTED`.

Reply:
done